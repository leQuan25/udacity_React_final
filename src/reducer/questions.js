import { RECEIVE_QUESTIONS, TOGGLE_QUESTIONS, ADD_QUESTIONS } from "../actions/questions";

export default function questions(state = {}, action) {
  switch (action.type) {
    case RECEIVE_QUESTIONS:
    return {
        ...state,
        questions: action.questions,
      };

      case TOGGLE_QUESTIONS:
      return {
        ...state,
        [action.qid]: {
            ...state.questions[action.qid],
            optionOne:
                action.answer === 'optionOne' ? state.questions[action.qid].optionOne.votes.concat([action.authedUser]) : state.questions[action.qid].votes,
            optionTwo:
                action.answer === 'optionTwo' ? state.questions[action.qid].optionTwo.votes.concat([action.authedUser]) : state.questions[action.qid].votes,
        },
      };

     case ADD_QUESTIONS:
        const { questions } = action;
        return {
            ...state,
            ...questions,
        };

    default:
      return state;
  }
}
