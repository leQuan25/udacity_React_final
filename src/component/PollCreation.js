import { useState } from "react";
import { connect } from "react-redux";
import { handleAddQuestion } from "../actions/questions";
import {Fragment } from "react";
import Nav from "./Nav";
import { useNavigate } from "react-router-dom";

const NewPoll = ({ dispatch }) => {
    const navigate = useNavigate();
    const [optionOne, setOptionOne] = useState("");
    const [optionTwo, setOptionTwo] = useState("");
  

    const handleChangeOptionOne = (e) => {
        const text = e.target.value;
        setOptionOne(text);
      };

    const handleChangeOptionTwo = (e) => {
        const text = e.target.value;
        setOptionTwo(text);
      };
   

      const handleSubmit = (e) => {
        e.preventDefault();
        dispatch(handleAddQuestion(optionOne,optionTwo));
        navigate("/home");
      };

    return (
        <Fragment>
            <Nav></Nav>
            <div>
                <h3>Would you rather</h3>
                <h1>Create yout own poll</h1>

                <form className="login" onSubmit={handleSubmit}>
                    <label htmlFor="optionOne">First option:</label>
                    <input 
                    type="text" 
                    id="optionOne" 
                    data-testid="optionOne"
                    name="optionOne" 
                    placeholder="Option One" 
                    value={optionOne}
                    onChange={handleChangeOptionOne} />
                    <br></br>
                    
                    <label htmlFor="optionTwo">Second option:</label>
                    <input 
                    type="text" 
                    id="optionTwo" 
                    data-testid="optionTwo"
                    name="optionTwo" 
                    placeholder="Option Two"
                    value={optionTwo}
                    onChange={handleChangeOptionTwo}
                    />
                    <br></br>
                    
                    <button data-testid="submitButton" className="btn" type="submit" disabled={optionOne === "" && optionTwo === ""}>
                    Submit
                    </button>
                </form>  
            </div>
        </Fragment>
    );
};

export default connect()(NewPoll);