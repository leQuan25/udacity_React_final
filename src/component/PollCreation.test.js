import {render, fireEvent} from '@testing-library/react';
import * as React from 'react';
import PollCreation from './PollCreation';
import reducer from '../reducer';
import middleware from '../middleware';
import {legacy_createStore as createStore} from 'redux';
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";

describe('Poll Creation', () => {
    it('will match snapshot', () => {
        const store = createStore(reducer, middleware);
        var component = render(<Provider store={store}><Router><PollCreation/></Router></Provider>);
        expect(component).toMatchSnapshot();
    });
});

describe('Poll Creation page display items check', () => {
 it('check disable and enable items', () => {
    const store = createStore(reducer, middleware);
        var component = render(<Provider store={store}><Router><PollCreation/></Router></Provider>);
        expect(component.getByTestId('submitButton')).toHaveAttribute('disabled');
        var optionOne = component.getByTestId('optionOne');
        var optionTwo = component.getByTestId('optionTwo');
        fireEvent.change(optionOne, {target: { value: 'optionOne'}});
        fireEvent.change(optionTwo, {target: { value: 'optionTwo'}});
        expect(component.getByTestId('submitButton')).toBeEnabled();
 });  
});
