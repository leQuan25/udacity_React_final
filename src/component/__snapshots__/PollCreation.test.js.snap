// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Poll Creation will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar"
      >
        <ul>
          <li
            class="navitem"
          >
            <a
              href="/home"
            >
              Home
            </a>
          </li>
          <li
            class="navitem"
          >
            <a
              href="/leaderboard"
            >
              Leaderboard
            </a>
          </li>
          <li
            class="navitem"
          >
            <a
              href="/new"
            >
              New
            </a>
          </li>
          <ul
            class="navbar-right"
          >
            <li
              class="navitem"
            />
            <li
              class="navitem"
            >
              <a
                href="/"
              >
                 Logout 
              </a>
            </li>
          </ul>
        </ul>
      </nav>
      <div>
        <h3>
          Would you rather
        </h3>
        <h1>
          Create yout own poll
        </h1>
        <form
          class="login"
        >
          <label
            for="optionOne"
          >
            First option:
          </label>
          <input
            data-testid="optionOne"
            id="optionOne"
            name="optionOne"
            placeholder="Option One"
            type="text"
            value=""
          />
          <br />
          <label
            for="optionTwo"
          >
            Second option:
          </label>
          <input
            data-testid="optionTwo"
            id="optionTwo"
            name="optionTwo"
            placeholder="Option Two"
            type="text"
            value=""
          />
          <br />
          <button
            class="btn"
            data-testid="submitButton"
            disabled=""
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar"
    >
      <ul>
        <li
          class="navitem"
        >
          <a
            href="/home"
          >
            Home
          </a>
        </li>
        <li
          class="navitem"
        >
          <a
            href="/leaderboard"
          >
            Leaderboard
          </a>
        </li>
        <li
          class="navitem"
        >
          <a
            href="/new"
          >
            New
          </a>
        </li>
        <ul
          class="navbar-right"
        >
          <li
            class="navitem"
          />
          <li
            class="navitem"
          >
            <a
              href="/"
            >
               Logout 
            </a>
          </li>
        </ul>
      </ul>
    </nav>
    <div>
      <h3>
        Would you rather
      </h3>
      <h1>
        Create yout own poll
      </h1>
      <form
        class="login"
      >
        <label
          for="optionOne"
        >
          First option:
        </label>
        <input
          data-testid="optionOne"
          id="optionOne"
          name="optionOne"
          placeholder="Option One"
          type="text"
          value=""
        />
        <br />
        <label
          for="optionTwo"
        >
          Second option:
        </label>
        <input
          data-testid="optionTwo"
          id="optionTwo"
          name="optionTwo"
          placeholder="Option Two"
          type="text"
          value=""
        />
        <br />
        <button
          class="btn"
          data-testid="submitButton"
          disabled=""
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
